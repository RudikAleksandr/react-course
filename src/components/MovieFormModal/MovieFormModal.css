
.modalFormMovie {
  backdrop-filter: blur(10px);
}

.modalFormMovie .modal-content {
  margin: auto;
  padding: 10px 40px;
  padding-bottom: 20px;
  background-color: #232323;
  width: 80%;
}

.modalFormMovie .modal-header {
  padding-bottom: 0;
}

.modalFormMovie .modal-header h2 {
  color: white;
  text-transform: uppercase;
  font-weight: 300;
}

.modalFormMovie .modal-header,
.modalFormMovie .modal-footer {
  border: none;
}

.modalFormMovie .close,
.modalFormMovie .close:hover {
  font-size: 2.4rem;
  color: white;
}

.modalFormMovie .form-group {
  margin-top: 25px;
}

.modalFormMovie .movieId {
  display: block;
  margin: 0;
  color: white;
  font-weight: normal;
}

.modal-footer .errorMessage {
  margin-right: 20px;
  color: #f65261;
  text-transform: uppercase;
  font-weight: bold;
}


