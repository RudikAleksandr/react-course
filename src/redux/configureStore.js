import { configureStore as configureStoreToolkit } from '@reduxjs/toolkit';
import thunkMiddleware from 'redux-thunk';
import moviesReducer from './moviesSlice';

const configureStore = (preloadedState) => (
  configureStoreToolkit({
    preloadedState,
    reducer: {
      movies: moviesReducer,
    },
    middleware: (getDefaultMiddleware) => getDefaultMiddleware().concat(thunkMiddleware),
  })
);

export default configureStore;
