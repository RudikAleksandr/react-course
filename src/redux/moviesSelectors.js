
const selectListMovies = (state) => state.movies.moviesOrder.map(id => state.movies.movies[id]);

const selectSearch = (state) => state.movies.search;

const selectActiveMovieGenre = (state) => state.movies.activeMovieGenre;

const selectActiveSortType = (state) => state.movies.activeSortType;

export {
  selectListMovies,
  selectSearch,
  selectActiveMovieGenre,
  selectActiveSortType
};