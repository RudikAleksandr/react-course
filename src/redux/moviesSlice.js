import { createAsyncThunk, createSlice } from '@reduxjs/toolkit';
import * as moviesAPI from '../api/moviesAPI';
import { groupByKey } from '../utils/grouping';

export const initialState = {
  movies: {},
  moviesOrder: [],
  search: '',
  activeMovieGenre: '',
  activeSortType: 'release_date'
};

export const getListMovies = createAsyncThunk(
  'movies/getListMovies',
  moviesAPI.getListMovies
);

export const getMovie = createAsyncThunk(
  'movies/getMovie',
  moviesAPI.getMovie
);

export const updateMovie = createAsyncThunk(
  'movies/updateMovie',
  moviesAPI.updateMovie
);

export const createMovie = createAsyncThunk(
  'movies/createMovie',
  moviesAPI.createMovie
);

export const deleteMovie = createAsyncThunk(
  'movies/deleteMovie',
  moviesAPI.deleteMovie
);

const moviesSlice = createSlice({
  name: 'movies',
  initialState,
  reducers: {
    setSearch: (state, { payload }) => {
      state.search = payload;
    },
    setActiveMovieGenre: (state, { payload }) => {
      state.activeMovieGenre = payload;
    },
    setActiveSortType: (state, { payload }) => {
      state.activeSortType = payload;
    }
  },
  extraReducers: {
    [getListMovies.fulfilled]: (state, { payload }) => {
      state.movies = groupByKey('id', payload.data);
      state.moviesOrder = payload.data.map((movie) => movie.id);
    },
    [updateMovie.fulfilled]: (state, { payload }) => {
      state.movies[payload.id] = payload;
    },
  },
});

export const {
  setSearch,
  setActiveMovieGenre,
  setActiveSortType
} = moviesSlice.actions;

export default moviesSlice.reducer;