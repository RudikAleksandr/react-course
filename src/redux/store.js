import { configureStore } from '@reduxjs/toolkit';
import moviesReducer from './moviesSlice';
import thunkMiddleware from 'redux-thunk'

const store = configureStore({
  reducer: {
    movies: moviesReducer,
  },
  middleware: (getDefaultMiddleware) => getDefaultMiddleware().concat(thunkMiddleware),
});

export default store;