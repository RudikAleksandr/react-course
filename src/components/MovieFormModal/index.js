import MovieFormModal from './MovieFormModal';

export default MovieFormModal;
