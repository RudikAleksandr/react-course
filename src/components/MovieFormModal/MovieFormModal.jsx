import React, { useMemo } from 'react';
import { Modal, Button } from 'react-bootstrap';
import { useDispatch } from 'react-redux';
import { useFormik } from 'formik';
import * as yup from 'yup';
import { createMovie, updateMovie } from '../../redux/moviesSlice';
import { FieldInput, FieldDate, FieldSelect, Label } from '../common';
import './MovieFormModal.css';

const MovieSchema = yup.object().shape({
  title: yup.string().required('(required)'),
  tagline: yup.string().required('(required)'),
  release_date: yup.date(),
  poster_path: yup.string().url('(invalid url)').required('(required)'),
  overview: yup.string().required('(required)'),
  runtime: yup.number('(required number)').typeError('(must be a number)').min(0, '(min 0)').required('(required)'),
  genres: yup.array().min(1, '(must have at least one genre)').required('(required)'),
});

const optionsGenres = [
  { value: 'Documentary', label: 'Documentary' },
  { value: 'Comedy', label: 'Comedy' },
  { value: 'Horror', label: 'Horror' },
  { value: 'Crime', label: 'Crime' }
];

const initialMovie = {
  title: '',
  tagline: '',
  release_date: new Date(),
  poster_path: '',
  overview: '',
  runtime: '',
  genres: []
};

const MovieFormModal = ({ movie = initialMovie, onHide }) => {
  const dispatch = useDispatch();
  const isCreateMode = useMemo(() => !movie.id, [movie.id]);
  const formik = useFormik({
    initialValues: movie,
    validationSchema: MovieSchema,
    onSubmit: (values) => {
      const actionMovie = isCreateMode ? createMovie : updateMovie;
      dispatch(actionMovie(values));
      onHide();
    },
    validateOnMount: true,
  });

  return (
    <Modal
      centered
      size="lg"
      show={true}
      className="modalFormMovie"
      onHide={onHide}
      data-testid="movieFormModal">
      <Modal.Header closeButton>
        <h2>{isCreateMode ? 'Add' : 'Edit'} movie</h2>
      </Modal.Header>
      <Modal.Body>
        {!isCreateMode &&
          <>
            <Label text="movie id" />
            <Label className="movieId" text={movie.id} />
          </>
        }
        <FieldInput
          label={`title ${formik.errors.title || ''}`}
          name="title"
          value={formik.values.title}
          placeholder="Input title"
          onChange={formik.handleChange}
        />
        <FieldInput
          label={`tagline ${formik.errors.tagline || ''}`}
          name="tagline"
          value={formik.values.tagline}
          placeholder="Input tagline"
          onChange={formik.handleChange}
        />
        <FieldDate
          label="release date"
          name="release_date"
          value={formik.values.release_date}
          placeholderText="Select release date"
          onChange={formik.setFieldValue}
        />
        <FieldInput
          label={`movie url ${formik.errors.poster_path || ''}`}
          type="url"
          name="poster_path"
          value={formik.values.poster_path}
          onChange={formik.handleChange}
          placeholder="Input URL"
        />
        <FieldSelect
          label={`genres ${formik.errors.genres || ''}`}
          name="genres"
          value={formik.values.genres}
          onChange={formik.setFieldValue}
          options={optionsGenres}
          placeholder="Select genre"
        />
        <FieldInput
          label={`overview ${formik.errors.overview || ''}`}
          name="overview"
          value={formik.values.overview}
          placeholder="Input overview"
          onChange={formik.handleChange}
        />
        <FieldInput
          label={`runtime ${formik.errors.runtime || ''}`}
          type="number"
          name="runtime"
          value={formik.values.runtime}
          placeholder="Input runtime"
          onChange={formik.handleChange}
        />
      </Modal.Body>
      <Modal.Footer>
        <Button onClick={formik.handleReset} variant="secondary">Reset</Button>
        <Button onClick={formik.handleSubmit} className="ml-3">{isCreateMode ? 'Submit' : 'Save'}</Button>
      </Modal.Footer>
    </Modal>
  );
}

export default MovieFormModal;
