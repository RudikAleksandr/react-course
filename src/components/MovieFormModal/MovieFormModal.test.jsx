import React from 'react';
import { render } from '@testing-library/react';
import { MemoryRouter, Route } from 'react-router';
import { Provider } from 'react-redux';
import configureStore from 'redux-mock-store';
import { initialState } from '../../redux/moviesSlice';

import MovieFormModal from './MovieFormModal';

jest.mock('../../redux/moviesSlice');

const getComponent = (store, props) => (
  <MemoryRouter>
    <Route>
      <Provider store={store}>
        <MovieFormModal movie={props.movie} onHide={props.onHide} />
      </Provider>
    </Route>
  </MemoryRouter>
);

const movie = {
  id: 'id',
  poster_path: '',
  title: 'title',
  release_date: new Date(2017, 1, 14),
  genres: [],
};

describe('MovieFormModal component', () => {
  test('render MovieFormModal snapshot', () => {
    const store = configureStore()({ movies: initialState });
    const component = getComponent(store, { movie, onHide: jest.fn() });
    const { asFragment } = render(component);

    expect(asFragment(component)).toMatchSnapshot();
  });

  // test('should call updateMovie', async () => {
  //   const store = configureStore()({ movies: initialState });
  //   jest.spyOn(reactRedux, 'useDispatch').mockReturnValue(jest.fn());

  //   const { getByRole } = render(getComponent(store, { movie, onHide: jest.fn() }));

  //   userEvent.click(getByRole('button', { name: /Save/i }));

  //   //     rerender(getComponent(store, { ...props, genre: 'all' }));

  //   expect(updateMovie).toHaveBeenCalled();

  //   await waitFor(() =>
  //     expect(updateMovie).toHaveBeenCalled()
  //   )
  // });
});
