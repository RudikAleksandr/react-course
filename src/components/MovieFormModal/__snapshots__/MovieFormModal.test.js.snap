// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MovieFormModal component render MovieFormModal snapshot 1`] = `<DocumentFragment />`;
